src/Composer/Util/Zip.php

```diff
@@ -84,28 +85,20 @@ private static function locateFile(\ZipArchive $zip, $filename)
             // handle archives with proper TOC
             if ($dirname === '.') {
                 $topLevelPaths[$name] = true;
-                if (\count($topLevelPaths) > 1) {
-                    // archive can only contain one top level directory
-                    return false;
-                }
-                continue;
+                throw new \RuntimeException('Archive has more than one top level directories, and no composer.json was found on the top level, so it\'s an invalid archive. Top level path found were: '.implode(',', array_keys($topLevelPaths)));
```

Comment:
This looks better but now it's going to throw always when it finds the first top level folder, which is not what we wanted. This is shown by the fact you had to adjust testReturnsComposerJsonInFirstFolder and testReturnsComposerJsonFromFirstSubfolder as they started throwing. The if above which is deleted should be restored, the throw should only be in place of return false, the rest of the logic was fine I think.

Reply:
ok reverted that back and added the test, I've also had to adjust the second condition and check that the dirname isn't a "." since you already check for that earlier